src/runtime/internal/memory_arena.h

```diff
@@ -0,0 +1,312 @@
+
+    // Default initial capacity
+    static constexpr uint32_t default_capacity = uint32_t(32);  // smallish
+
+    // Configurable parameters
+    struct Config {
+        uint32_t entry_size = 1;
+        uint32_t minimum_block_capacity = default_capacity;
+        uint32_t maximum_block_count = 0;
+    };
+
+    MemoryArena(void *user_context, const Config &config = default_config(),
+                const SystemMemoryAllocatorFns &allocator = default_allocator());
+
+    ~MemoryArena();
+
+    // Factory methods for creation / destruction
+    static MemoryArena *create(void *user_context, const Config &config, const SystemMemoryAllocatorFns &allocator = default_allocator());
+    static void destroy(void *user_context, MemoryArena *arena);
+
+    // Initialize a newly created instance
+    void initialize(void *user_context, const Config &config,
+                    const SystemMemoryAllocatorFns &allocator = default_allocator());
+
+    // Public interface methods
+    void *reserve(void *user_context, bool initialize = false);
+    void reclaim(void *user_context, void *ptr);
+    bool collect(void *user_context);  //< returns true if any blocks were removed
+    void destroy(void *user_context);
+
+    // Access methods
+    const Config &current_config() const;
+    static const Config &default_config();
+
+    const SystemMemoryAllocatorFns &current_allocator() const;
+    static const SystemMemoryAllocatorFns &default_allocator();
+
+private:
+    // Sentinal invalid entry value
+    static const uint32_t invalid_entry = uint32_t(-1);
+
+    // Each block contains:
+    // - an array of entries
+    // - an array of indices (for the free list)
+    // - an array of status flags (indicating usage)
+    // - free index points to next available entry for the block (or invalid_entry if block is full)
+    struct Block {
+        void *entries = nullptr;
+        uint32_t *indices = nullptr;
+        AllocationStatus *status = nullptr;
+        uint32_t capacity = 0;
+        uint32_t free_index = 0;
+    };
+
+    Block *create_block(void *user_context);
+    bool collect_block(void *user_context, Block *block);  //< returns true if any blocks were removed
+    void destroy_block(void *user_context, Block *block);
+    Block *lookup_block(void *user_context, uint32_t index);
+
+    void *create_entry(void *user_context, Block *block, uint32_t index);
+    void destroy_entry(void *user_context, Block *block, uint32_t index);
+    void *lookup_entry(void *user_context, Block *block, uint32_t index);
+
+    Config config;
+    BlockStorage blocks;
+};
+
+MemoryArena::MemoryArena(void *user_context,
+                         const Config &cfg,
+                         const SystemMemoryAllocatorFns &alloc)
+    : config(cfg),
+      blocks(user_context, {sizeof(MemoryArena::Block), 32, 32}, alloc) {
+    halide_debug_assert(user_context, config.minimum_block_capacity > 1);
+}
+
+MemoryArena::~MemoryArena() {
+    destroy(nullptr);
```

Comment:
This troubles me; either the code needs user_context to be safe (in which case we can't pass nullptr here), or it doesn't (in which case, why is passing it even an option)?